### ISSUE TITLE
The simulation of several models has slowed down significantly.

### ISSUE THREAD

**Issue by niklwors** [OpenModelica/OpenModelica#9131]
### Description

Due to changes in our BRSL Valves library, the simulation for a large part of the models has slowed down considerably. With the CVode solver, the simulation stops after a short time. With other Modelica tools, however, the simulation time has not changed, both with the Dassl and the CVode solver.
Here is a comparison of the times of BRSL 2.1 and BRSL 2.2 in the current state of development.
BRSL Valves 2.1 for the test model: Rexroth_Valves_Test.Valves_4WR.Rexroth_4WREE6E32_3X.VC43_DO_TC01
```
### STATISTICS ###
timer
    0.430551s          reading init.xml
     2.53206s [ 59.1%] simulation
     4.28611s [100.0%] total
   0.0694676s          reading info.xml
   0.0246807s [  0.6%] pre-initialization
    0.004187s [  0.1%] initialization
   0.0006492s [  0.0%] steps
   0.0701839s [  1.6%] solver (excl. callbacks)
     1.13172s [ 26.4%] creating output-file
    0.443842s [ 10.4%] event-handling
   0.0787901s [  1.8%] overhead
events
   17 state events
    0 time events
solver: dassl
11265 steps taken
19681 calls of functionODE
 3662 evaluations of jacobian
 1840 error test failures
    1 convergence test failures
0.130555s time of jacobian evaluation
The simulation finished successfully.
```
BRSL Valves 2.2 dev for the test model: Rexroth_Valves_Test.Valves_4WR.Rexroth_4WREE6E32_3X.VC43_DO_TC01
```
The initialization finished successfully without homotopy method.
### STATISTICS ###
timer
    0.594878s          reading init.xml
     229.023s [ 82.7%] simulation
     276.813s [100.0%] total
     0.14448s          reading info.xml
   0.0325003s [  0.0%] pre-initialization
   0.0106421s [  0.0%] initialization
   0.0118627s [  0.0%] steps
     18.6403s [  6.7%] solver (excl. callbacks)
      2.6204s [  0.9%] creating output-file
      26.341s [  9.5%] event-handling
     0.13386s [  0.0%] overhead
events
   19 state events
    0 time events
solver: dassl
5505151 steps taken
17495235 calls of functionODE
7940195 evaluations of jacobian
3978398 error test failures
    1 convergence test failures
157.552s time of jacobian evaluation
The simulation finished successfully.
```
### Steps to Reproduce
Load the BRSL package: Rexroth_Valves_Test\package.mo
Simulate the model: Rexroth_Valves_Test.Valves_4WR.Rexroth_4WREE6E32_3X.VC43_DO_TC01

### Version and OS

 - OpenModelica Version: v1.20.0-dev-153-gd60a541851 (64-bit)
 - OS:  Windows 10, 64 bit
 - MSL 4.0.0
 - BRSL 2.2 dev (June 21, 2022)

**Comment by casella** [comment]
OMG, 5.5 million integration steps with 3.9 million rejected steps. I guess you can check which derivative(s) in the simulation is chattering, maybe you are using noEvent on some expression that is actually discontinuous?

**Comment by niklwors** [comment]
@casella  @sbaetzing  The variable that is chattering now is Rexroth_Valves.Internal.Templates.Blocks.PT1Variable.der(y)
For the PT1Variable class, an MSL second order class was previously used instead. In the new class PT1Variable, T1 is a variable now  before it was a parameter.

**Comment by casella** [comment]
> @casella @sbaetzing The variable that is chattering now is Rexroth_Valves.Internal.Templates.Blocks.PT1Variable.der(y)

Very good, this allows to pinpoint the root cause easily.

> For the PT1Variable class, an MSL second order class was previously used instead. In the new class PT1Variable, T1 is a variable now before it was a parameter.

I don't have the model at hand, but I guess T1 is actually time-varying, possibly because of some gain-scheduling control. Is it also chattering? I'm also a bit confused, isn't a PT1 block meant to be a first-order system with transfer function mu/(1+sT1) ?

**Comment by niklwors** [comment]
Sorry, a first order component was used before.

**Comment by niklwors** [comment]
T1 is not chattering

**Comment by casella** [comment]
@niklwors I got the model from Adrian, I'll look at it ASAP, later this week if I manage.

**Comment by casella** [comment]
@niklwors, we had similar issues with the Buildings library, that were resolved by using the `--generateSymbolicJacobian` command line flag. Apparently, in some cases the numerical Jacobian is bad for some reason, and this causes a large number of convergences failures and rejected steps, but the issue is solved outright by using the analytic Jacobian.

Could you please try this out and report?

Thanks!

**Comment by casella** [comment]
@AnHeuermann, @phannebohm, @kabdelhak, then we have two choices to solve this issue:
1. try to understand why the numerical Jacobian got worse at some point
2. try to get Niklas' model to work with symbolic Jacobians

Which one would you pick?

**Comment by casella** [comment]
If SymbolicJacobian.deriveAll fails just because of one (or a few) function(s) that cannot be symbolically differentiated, we should use numerical differentiation for them. Apparently there is some code around to do that, but it's not used. We could consider doing that at this point, rather than falling back to full numerical jacobian.

**Comment by kabdelhak** [comment]
We fixed the differentiation of the delay operator in PR #9343 which allowed us to compare the symbolical and numerical simulation time. It seems like there is not much of a difference so the problem should be somewhere else, maybe the sparsity pattern is faulty. 

Could you provide the library in two different versions?
 - the version where `Rexroth_Valves_Test.Valves_4WR.Rexroth_4WREE6E32_3X.VC43_DO_TC01` still was fast
 - the version where `Rexroth_Valves_Test.Valves_4WR.Rexroth_4WREE6E32_3X.VC43_DO_TC01` first was slow 
 
It would help us a lot to pin down the exact problem and not lose time following red herrings.

**Comment by niklwors** [comment]
@kabdelhak I have tested this model in the version mentioned in the ticket with the current OpenModelica version and now the model simulates fast again. The numerical Jacobian matrix calculation is set. (As far as I have seen it correctly, Dymola also uses the numerical Jacobian matrix by default) I don't know then what the problem was. But I will start our complete test suite again and give feedback if it occurs again.

**Comment by casella** [comment]
All's well that ends well 😅 .

@niklwors if the outcome of the testsuite is positive, please close this ticket. Thanks!

**Comment by niklwors** [comment]
Unfortunately, there are still models that simulate very slowly. For example, the test model.  
```Rexroth_Valves_Test.Valves_5WR.Rexroth_5WRPE10F1B70L_2X.VC53_DO_TC01```
In version 2.1 these valve models did not yet exist
```
### STATISTICS ###
timer
	0.248131s          reading init.xml
	 250.114s [ 82.7%] simulation
	 302.559s [100.0%] total
   0.0366889s          reading info.xml
   0.0032132s [  0.0%] pre-initialization
   0.0036361s [  0.0%] initialization
	0.001254s [  0.0%] steps
	 19.5134s [  6.4%] solver (excl. callbacks)
	 6.46613s [  2.1%] creating output-file
	 26.1145s [  8.6%] event-handling
	0.343015s [  0.1%] overhead
events
   18 state events
	0 time events
solver: dassl
9359266 steps taken
28599347 calls of functionODE
14483285 evaluations of jacobian
6541281 error test failures
	0 convergence test failures
173.1s time of jacobian evaluation

```
But it is the same problem, the solver needs very many steps and rejects very many.

**Comment by casella** [comment]
Looks like the usual issue. We'll look into that.

**Comment by niklwors** [comment]
There are also still a view models that now simulate slowly and before with BRSL 2.1.1 it was fast. E.g. the model Rexroth_Valves_Test.Valves_4WR.Rexroth_4WSE2ED10_5X_5.VC43_PO_TC11a. I will provide you with a  current version of the BRSL v2.2 and  v2.1.1.
```
### STATISTICS ###
timer
    0.251586s          reading init.xml
     300.387s [ 81.6%] simulation
     368.113s [100.0%] total
   0.0390606s          reading info.xml
   0.0184163s [  0.0%] pre-initialization
   0.0041692s [  0.0%] initialization
   0.0001045s [  0.0%] steps
     33.3584s [  9.1%] solver (excl. callbacks)
     1.08479s [  0.3%] creating output-file
     33.2095s [  9.0%] event-handling
   0.0504988s [  0.0%] overhead
events
solver: dassl
16840868 steps taken
52138284 calls of functionODE
25396216 evaluations of jacobian
11799778 error test failures
   25 convergence test failures
216.1s time of jacobian evaluation

```

**Comment by casella** [comment]
Wow, 16 million steps and 52 million RHS computations is a record 😓

**Comment by casella** [comment]
I carried out some analysis, comparing the simulation performance between omc and Dymola, with equidistant time grid turned off to see the bare integrator performance.

Until time = 0.5, the performance is comparable. After the event at time = 0.5, which causes the increase of Valve.valve.valveDynamics_Complex.valveDynamics.commandSignalSwitchOff.limIntegrator.y, the number of steps taken by dassl in omc increases dramatically, compared to Dymola:

| StopTime  | Steps OMC | Jacobians OMC | Steps Dymola | Jacobians Dymola 
| ------------- | ------------- | ------------- | ------------- | ------------- | 
| 0.4 | 513  |  99 | 479 | 96 |
| 0.5 | 878 | 153 | 480 | 96 |
| 0.51 | 3616 | 3942 | 638 | 137 |
| 0.6 | 19253 | 22041 | 663 | 144 |
| 1.0 | 96704 | 115157 | 673 | 147 |

It is pretty obvious that after the event is triggered, something goes badly wrong with the Jacobian. The integration algorithm is able to trudge along somehow, but as soon as h is increased a bit, (I - hJ)^-1 is so badly wrong that convergence of the BDF formula is apparently hindered, triggering an outrageus number of Jacobian recomputations, which seems to be totally unjustified

**Comment by casella** [comment]
Some more information: I turned on LOG_SOLVER and LOG_JAC. Here is an excerpt of the log:
```
call solver from 0.500605 to 0.51 (stepSize: 0.00939499994981574)
DASSL-Solver: Matrix A
    -10000          0          0          0          0          0 
3.82964e+06   -2109.57          0 -3.82964e+06         -0          0 
         0          0          0          0          0          0 
         0          1          0          0          0          0 
         0          0          0 3.72705e+06   -109.019 5.43206e+06 
         0          0          0          0          0      -1000 
DASSL-Solver: Matrix A
    -10000          0          0          0          0          0 
3.82964e+06   -2109.58          0 -3.82964e+06         -0          0 
         0          0          0          0          0          0 
         0          1          0          0          0          0 
         0          0          0     106487   -109.019 5.43206e+06 
         0          0          0          0          0      -1000 
DASSL-Solver: Matrix A
    -10000          0          0          0          0          0 
3.82964e+06    -2109.6          0 -3.82964e+06         -0          0 
         0          0          0          0          0          0 
         0          1          0          0          0          0 
         0          0          0          0   -109.019 5.43206e+06 
         0          0          0          0          0      -1000 
DASSL-Solver: Matrix A
    -10000          0          0          0          0          0 
3.82964e+06   -2109.58          0 -3.82964e+06         -0          0 
         0          0          0          0          0          0 
         0          1          0          0          0          0 
         0          0          0          0   -109.019 5.43206e+06 
         0          0          0          0          0      -1000 
DASSL-Solver: Matrix A
    -10000          0          0          0          0          0 
3.82964e+06   -2109.59          0 -3.82964e+06         -0          0 
         0          0          0          0          0          0 
         0          1          0          0          0          0 
         0          0          0   -53243.6   -109.019 5.43206e+06 
         0          0          0          0          0      -1000 
DASSL-Solver: Matrix A
    -10000          0          0          0          0          0 
3.82964e+06   -2109.59          0 -3.82964e+06         -0          0 
         0          0          0          0          0          0 
         0          1          0          0          0          0 
         0          0          0          0   -109.019 5.43206e+06 
         0          0          0          0          0      -1000 
```
Here it is apparent where the problem is: element (5,4) of the Jacobian gets the following sequence of variables, each time the Jacobian is recomputed:
```
3.72705e+06
106487
0
0
-53243.6
0
-1.4624e+06
26998.9
-879941
```
while all other terms are basically constant. So, there's something wrong in how the fourth state variable affects the derivative of the fifth one. 

The list of states I get from LOG_SOLVER_V is:
```
1: Valve.valve.valveDynamics_Complex.valveDynamics.PT1.y = -1 (pre: -1)
2: Valve.valve.valveDynamics_Complex.valveDynamics.a_to_v.y = 0 (pre: 0)
3: Valve.valve.valveDynamics_Complex.valveDynamics.commandSignalSwitchOff.limIntegrator.y = 1 (pre: 1)
4: Valve.valve.valveDynamics_Complex.valveDynamics.v_to_s.y = -1 (pre: -1)
5: initialization.volume.variableVolume.p = 100000 (pre: 100000)
6: testSignal.firstOrder.y = 2.75 (pre: 2.75)
```
so, apparently, something's wrong in how valveDynamics.v_to_s.y influences the derivative of variableVolume.p. Upon inspection of the model, it seems that valveDynamics.v_to_s.y goes straight into the output of valveDynamics_Complex, which gets into the input of the four ControlEdge models, in particular ControlEdgeModelPA, which ultimately determines the flow out of initialization.volume, hence the derivative of its pressure.

So, there is something going bad in the signal path between the ControlEdgeModelPA input and the flow rate from initializazion.volume into port A of the valve.

I had a quick look to that and tried to figure out what could be going wrong, in terms of strong nonlinearities. First you encounter a limiter, which is not involved because the input signal is close to zero and the limits are +/- 1.1. Then there is a Table1D, which is also most likely not the culprit, since it works in a linear regime between time = 0.5 and time = 0.6. Then the signal enters the Resistor model, which is most likely the culprit, as it contains several if statements, some with noEvent.

Most likely, the critical equation is the following one, that is found in Rexroth_Valves.Internal.Templates.Components.GenericResistor_Complex:
```
  if Ahyd < EPS or abs(dp) < EPS or Modelica.Math.isEqual(
              c1Int,
              0.0,
              EPS) and Modelica.Math.isEqual(
              c2Int,
              0.0,
              EPS) then
    q = 0;
  elseif Modelica.Math.isEqual(
              c1Int,
              0.0,
              EPS) and not Modelica.Math.isEqual(
              c2Int,
              0.0,
              EPS) then
    q = (sqrt(abs(dp)/c2Int)*x)*sign(dp);
  elseif Modelica.Math.isEqual(
              c2Int,
              0.0,
              EPS) and not Modelica.Math.isEqual(
              c1Int,
              0.0,
              EPS) then
    q = (abs(dp)/c1Int*x^2)*sign(dp);
  else
    q = ((sqrt(c1Int^2 + 4.0*c2Int*abs(dp)*x^2) - c1Int)/2.0/c2Int)*sign(dp);
  end if;
```
Based on the LOG_SOLVER_V diagnostics:
```
Using Jacobian method: Colored numerical Jacobian.
```
my educated guess is that events in those if's are not handled correctly when computing the finite differences for the coloured Jacobian. At the moment I couldn't say whether it is a wrong scaling of the used delta's, or that the if conditions are taken literally (as if they had a noEvent, instead of always staying on the current branch), or if there is something wrong with sqrt, abs and sign, always regarding event handling. 

BTW, abs and sign are supposed not to generate events, and to be taken literally, see [Section 3.7.1](https://specification.modelica.org/maint/3.5/operators-and-expressions.html#numeric-functions-and-conversion-functions) of the MLS.

@AnHeuermann, @phannebohm, based on this intel, could you please go on with the debugging? I guess this problem is affecting _a lot_ of other models, also in the Buildings library, so I think it is of the utmost importance to fix it.

**Comment by casella** [comment]
FYI, the critical flow rate that determines the swinging Jacobian element is most likely this one:
```
(assign) Valve.qPA := if Valve.valve.controlEdgeModelPA.Resistor.Ahyd < 1e-15 or abs(dpPB) < 1e-15 or noEvent(abs(Valve.valve.controlEdgeModelPA.Resistor.c1Int) <= 1e-15) and noEvent(abs(Valve.valve.controlEdgeModelPA.Resistor.c2Int) <= 1e-15) then 0.0 else if noEvent(abs(Valve.valve.controlEdgeModelPA.Resistor.c1Int) <= 1e-15) and not noEvent(abs(Valve.valve.controlEdgeModelPA.Resistor.c2Int) <= 1e-15) then sqrt(abs(dpPB) / Valve.valve.controlEdgeModelPA.Resistor.c2Int) * Valve.valve.controlEdgeModelPA.Ahyd * Valve.valve.controlEdgeModelPA.Resistor.signdp else if noEvent(abs(Valve.valve.controlEdgeModelPA.Resistor.c2Int) <= 1e-15) and not noEvent(abs(Valve.valve.controlEdgeModelPA.Resistor.c1Int) <= 1e-15) then abs(dpPB) * Valve.valve.controlEdgeModelPA.Ahyd ^ 2.0 * Valve.valve.controlEdgeModelPA.Resistor.signdp / Valve.valve.controlEdgeModelPA.Resistor.c1Int else 0.5 * Valve.valve.controlEdgeModelPA.Resistor.signdp * (sqrt(Valve.valve.controlEdgeModelPA.Resistor.c1Int ^ 2.0 + 4.0 * Valve.valve.controlEdgeModelPA.Resistor.c2Int * abs(dpPB) * Valve.valve.controlEdgeModelPA.Ahyd ^ 2.0) - Valve.valve.controlEdgeModelPA.Resistor.c1Int) / Valve.valve.controlEdgeModelPA.Resistor.c2Int
```
I checked the values of c1Int and c2Int during the simulation, which are about 4e9 and 4e14 so I guess the conditions noEvent(abs(Valve.valve.controlEdgeModelPA.Resistor.c2Int) <= 1e-15) and noEvent(abs(Valve.valve.controlEdgeModelPA.Resistor.c2Int) <= 1e-15) can be assumed to be false, while dpPB is between 1e5 and 2e5.

I didn't have the heart to check the actual generated C-code, which may have additional quirks.

@niklwors I am wondering what sense does it make to check that c2Int is less than 1e-15 if it normally has order of magnitude 1e14. This model seems really a bit fishy to me, or at least not too easy to handle numerically.

**Comment by sbaetzing** [comment]
@casella, we are open to any model improvements aswell. We know that class (Rexroth_Valves.Internal.Templates.Components.GenericResistor_Complex) is a little bit tricky. Approach comes from a non equation based simulation environment.

thank you for the analysis. We will also take a look at the implemented equation.

**Comment by casella** [comment]
> @casella, we are open to any model improvements aswell. We know that class (Rexroth_Valves.Internal.Templates.Components.GenericResistor_Complex) is a little bit tricky. Approach comes from a non equation based simulation environment.

In fact, from the point of view of an object-oriented modelling fan, there are several things in that model that I found a bit questionable, but it was pretty obvious to me that they were just ported from an old-fashioned simulation code, and of course one has to make compromises in that case.

In the meantime I thought about a couple more experiments that I can perform without too much effort to try to pin down the root cause of this ticket. I'll carry them out ASAP and report here.

**Comment by casella** [comment]
I went on with the analysis with @mahge.

First of all, I removed the big if-equation in the Rexroth_Valves.Internal.Templates.Components.GenericResistor_Complex and replaced it with the equation actually holds during that transient: 
```
q = ((sqrt(c1Int^2 + 4.0*c2Int*abs(dp)*x^2) - c1Int)/2.0/c2Int)*sign(dp);
```
The simulation results were the same as before (and also the same as in Dymola), and the number of steps still went through the roof after time = 0.5, as before. Hence, the if-equation is not to blame, so we can get rid of it to ease the debugging effort.

Next, I got worried that the equation shown above was numerically ill-conditioned, because of the subtraction between two very large numbers. The equation for `q` is obviously the solution of a quadratic equation (which I myself would leave as an implicit one in Modelica, but that is another story for @niklwors). 

As it is well-know, see e.g. [here](https://people.csail.mit.edu/bkph/articles/[Quadratics.pdf](https://people.csail.mit.edu/bkph/articles/Quadratics.pdf)), when a quadratic equation has two roots with widely different orders of magnitude, the textbook formula `x = (-b + sqrt(b^2 - 4ac))/2a` (assuming b>0) is numerically ill-conditioned, because it takes the difference of two large numbers which are nearly equal. In this case, it is better to use the alternative formula x = 2c/(-b - sqrt(b^2 - 4ac)), which is equivalent, but avoids taking the ill-conditioned difference.

By using that formula, the equation for q can be rewritten as
```
q = 2*abs(dp)*x^2/(c1Int + sqrt(c1Int^2 + 4.0*c2Int*abs(dp)*x^2));
```
or even better as 
```
q = 2*dp*x^2/(c1Int + sqrt(c1Int^2 + 4.0*c2Int*abs(dp)*x^2));
```
so that it also would work for negative dp. But that doesn't matter, because dp >> 0 in this experiment. Once again, this clearly shows that porting non-equation-based models into Modelica is sometimes a sub-optimal choice; if the model contained the original implicit equation, the tool would have known how to deal with that automatically. But as we said, this is another story.

Anyway, I re-ran the model with the rewritten equation, but I got the same results and also the same huge number of steps, so that is not the culprit as well.

**Comment by casella** [comment]
## Recap
These are the facts I found out so far:
- The test model is `Rexroth_Valves_Test.Valves_4WR.Rexroth_4WSE2ED10_5X_5.VC43_PO_TC11a`
- You can simplify the analysis by replacing the big if-equation in `Rexroth_Valves.Internal.Templates.Components.GenericResistor_Complex` with  `q = 2*dp*x^2/(c1Int + sqrt(c1Int^2 + 4.0*c2Int*abs(dp)*x^2));`. This doesn't change the outcome of the experiment, while getting rid of a nasty if-expression in the critical dependency chain.
- Run the model with equidistant time grid turned off, protected variables output turned on, and StopTime = 0.6.
- The solver takes about 20.000 steps (almost all of them after time = 0.5), and about as many Jacobian recomputations. A reasonable number, based of Dymola's output, should be about 1000 steps and about 100 Jacobian recomputations.
- This means that there is something badly wrong with the numerically computed Jacobian, that prevents the step size to increase significantly without hampering the convergence of the BDF solver
- Turning on `LOG_SOLVER` and `LOG_JAC` reveals that there is some problem with the element (5,4) of the Jacobian, which oscillates wildly after each step, when time > 0.5. All other non-zero elements remain practically constant.
- This element is the sensitivity of the derivative of `initialization.volume.variableVolume.p` to changes of `Valve.valve.valveDynamics_Complex.valveDynamics.v_to_s.y`
- `der(initialization.volume.variableVolume.p)` is related to the flow rate `Valve.valve.controlEdgeModelPA.Resistor.q`, a.k.a. `Valve.qPA`, by the capacitance of `initialization.volume`. This flow rate is determined by the `Resistor` component on the basis of its `x` input, which comes from the output of the integrator `Valve.valve.valveDynamics_Complex.valveDynamics.v_to_s.y`, which is processed by a few intermediate blocks, including one Table1D block
- as I understand based on the output of the simulation, those intermediate blocks have a linear behaviour in this simulation, so they should not be involved in the issue

At the end of the day, something goes wrong in the computation of the chain of dependencies from `Valve.valve.valveDynamics_Complex.valveDynamics.v_to_s.y` to `der(initialization.volume.variableVolume.p)` when computing numerical coloured Jacobians. I guess this is as far as I can get with my modelling knowledge, from this point onwards it is up to  @AnHeuermann and @mahge to figure out what is going wrong there.

One possible cause is that there is some numerical issue, e.g. too small delta-x, that makes the (5,4) element critically dependent on rounding errors. Another possibility is that there is something wrong in the sparsity pattern detection, so there are some variables in the generated code that are read before they are written, or something like that. Maybe the pattern of the values of element (5,4) reported above gives you some hint. As I understand, their order of magnitude is reasonable, though they should not get to zero, and their sign should remain the same all the time (not sure if positive or negative).

Identifying the root cause of this bug and fixing it is a top-priority issue, not only for BR. In fact, there are many other models failing similarly in other libraries, so this is not specific to BR's libraries. For example, [Buildings_latest_Buildings.Examples.DualFanDualDuct.ClosedLoop](https://libraries.openmodelica.org/branches/master/Buildings_latest/files/Buildings_latest_Buildings.Examples.DualFanDualDuct.ClosedLoop.sim) has the following statistics:
```
|                 | |       | | solver: dassl
|                 | |       | | | 61888 steps taken
|                 | |       | | | 136011 calls of functionODE
|                 | |       | | | 87030 evaluations of jacobian
|                 | |       | | |   215 error test failures
|                 | |       | | | 28724 convergence test failures
|                 | |       | | | 225.281s time of jacobian evaluation
```
The pattern is exactly the same: the simulation is horrendously slow because the solver takes a huge number of small steps, and triggers an insane number of Jacobian recomputations regardless. It seems to me that assuming the same root cause is a safe bet. I'm not sure if I should spend another 6 hrs or so of analysis work on this other model, to figure out where the problem is, particularly since this other model has 224 state variables, so debugging the Jacobian will be a big deal.

In fact, I think we're lucky that this BR test case is not too large, so we can use it for debugging.

@AnHeuermann, @mahge, good luck with that and let me know if you need more insight 😄

**Comment by AnHeuermann** [comment]
I thought it would be interesting to plot the step size around `t=0.5` so here it is:

![grafik](https://user-images.githubusercontent.com/38031952/201646473-212f77f9-b234-47b3-9ea8-0c9fa5d18810.png)
![grafik](https://user-images.githubusercontent.com/38031952/201646563-6778c454-742f-4274-a845-67ce5c751393.png)

Maybe we should remove all `noEvent`'s and see how many events the ODE integrator has to handle after `t=0.5`.

**Comment by AnHeuermann** [comment]
Also the result file is like 139GB 💥 on my Windows system (20GB on my Linux). No wonder this is slow...

**Comment by AnHeuermann** [comment]
We have sparsity pattern

```
          x1 x2 x4 x3 x6 x5
der(x1)   *               
der(x2)   *  *     *  *  *
der(x4)         *         
der(x5)      *     *      
der(x6)            *  *  *
der(x5)                  *
```
with

```
x1 := Valve.valve.valveDynamics_Complex.valveDynamics.PT1.y,                                  --> State 1
x2 := Valve.valve.valveDynamics_Complex.valveDynamics.a_to_v.y,                               --> State 2
x3 := Valve.valve.valveDynamics_Complex.valveDynamics.v_to_s.y,                               --> State 4
x4 := Valve.valve.valveDynamics_Complex.valveDynamics.commandSignalSwitchOff.limIntegrator.y  --> State 3
x5 := testSignal.firstOrder.y,                                                                --> State 6
x6 := initialization.volume.variableVolume.p}                                                 --> State 5
```

It turns out, that `der(x4)` doesn't depend on any state. So we have a zero row/column in our Jacobian.
The reason it is shown in the sparsity pattern is, that inside the if condition of `equation 55` `x4` is used, but not in the assignment itself. In the end we have something like

```modelica
der(x4) := if(condition) then -constValue else constValue end if;
```

@kabdelhak is trying to fix the dependency analysis to ignore variables inside the conditions (if, when, ...).

**Comment by AnHeuermann** [comment]
@kabdelhak and myself created a minimal example that behaves somewhat similar to `VC43_PO_TC11a`. It is super slow and has `noEvent`.

```modelica
model test_min
  Real x(start=1.0, fixed=true),y(start=0.0, fixed=true);
  parameter Real a=-0.01; // Fine tune to get OMC as slow as needed.
equation
  der(x) = x*(y+1);
  der(y) = if noEvent(y > 0.5) then a else 1.0;
end test_min;
```

Running this in Dymola produces a warning
```
Warning message from dymosim
At time T = 1.046536e+00 in current integration interval
T_interval = 1.045000e+00 ... 1.047200e+00
a large amount of work has been expended
(about 500 steps) in the integrator. Probably the communication
interval is too large or the system is stiff.
```
but maybe you can specify the parameter `a` to be below the threshold of Dymola but still be slower in OMC.
Of course the original is slightly more complicated, but maybe this is the root issue?

We are not sure what the sparsity pattern should be in this case. The Jacobian
```
* 0
* 0
```
is singular.

**Comment by casella** [comment]
> The Jacobian is singular

This is not an issue, the BDF algorithm needs to invert/factorize (I - h*J), which is always non-singular for sufficiently small h even if J is

> It turns out, that `der(x4)` doesn't depend on any state.

This is absolutely normal, since x4 is the state variable of an integral controller, whose equation (without considering the limits) is just `der(x) = u`.

**Comment by casella** [comment]
> @kabdelhak is trying to fix the dependency analysis to ignore variables inside the conditions (if, when, ...).

I am not 100% sure this is always a good idea. It depends whether or not events are generated. Consider this MWE:
```
model M
  Real x1(start = 0, fixed = true);
  Real x2(start = 0, fixed = true);
equation
  der(x1) = 2;
  der(x2) = noEvent(if x1 > 1 then 1 else 0);
end M;
```
In this case ODE integration is performed without event handling as `x1` passes the threshold `x1 = 1`. If you want the RHS to be computed correctly at each time step during continuous time integration, you have to take into account the fact that the derivative of x2 actually depends on x1, even though this dependency is only in some conditions.

If you instead consider
```
model M
  Real x1(start = 0, fixed = true);
  Real x2(start = 0, fixed = true);
equation
  der(x1) = 2;
  der(x2) = if x1 > 1 then 1 else 0;
end M;
```
during continuous time integration the if expression is locked onto a branch, in this case `der(x2) = 0`, and should remain so even as the zero crossing function changes sign, and during the iterations to find the exact event time. In this case, it is correct to neglect the influence of state variables only appearing in conditions. 

Summing up, the dependencies of conditions on the RHS of the ODEs should only be considered when they are involved in expressions that do not generate events, because of noEvent, smooth, or because operators that do not generate events, like abs() or sign().

That said, I don't see any big problem if extra dependencies are marked in the sparsity pattern - if the derivatives are computed correctly, the only consequence will be to waste some time computing zero terms that could have been skipped in the first place, and wasting some more time in the factorization of the (I - h*J) matrix. This is only a matter of efficiency, not of correctness of the computed Jacobian.

**Comment by casella** [comment]
> We are not sure what the sparsity pattern should be in this case. The Jacobian
> 
> ```
> * 0
> * 0
> ```

I am confused. Rows should correspond to explicit differential equations and columns to states, right? It seems to me you are showing the transposed sparsity pattern, is that the case. As I understand, given the noEvent and what I wrote previously, the sparsity pattern should be

|    |  x | y |
|--|----|---|
|eq1 | * | * |
|eq2 | 0 | * | 

That said, the `test_min` system doesn't have a solution in the classical sense of Cauchy, because it has a sliding mode y = 0.5 where the solution is only defined in [Filippov's](https://en.wikipedia.org/wiki/Differential_inclusion) generalized sense, so this would be problematic for any Modelica tool. If events are activated, you get a ton of events. If not, you get a ton of small steps, but the final outcome is the same: this situation should be avoided. Unless you use special symbolic handling for sliding modes, see e.g. [Siconos](https://en.wikipedia.org/wiki/Siconos).

**Comment by casella** [comment]
Getting back to our problem, I think your MWE is interesting but not necessarily captures the original issue. If I simulate it with no equidistant time grid (so we see the naked variable step-size ODE solver behaviour) with stopTime = 0.52 and Tolerance = 1e-4 and plot der(y) using Dymola, I get what I would expect:
![immagine](https://user-images.githubusercontent.com/4233148/202140297-21fa5e85-1417-468c-97bb-36a38ed41692.png)
the derivative oscillates at very high frequency (8636 steps are taken) between a positive and a negative value, keeping y constant at 0.5. Not nice, but makes sense.

If I do the same with OMC, it takes 3497 steps and I get this:
![immagine](https://user-images.githubusercontent.com/4233148/202141248-d429b777-40d8-4028-b001-1ea5ac021f48.png)
which makes no sense at all, because it shows me a constant derivative of -0.01 after time = 0.5, despite the fact that y remains constant. You may want to investigate why the steps with positive derivative are not reported, since for sure they have to be there, but this is probably another story. Maybe.

The point is, if I check the derivatives of the BR model with OMC, I get this behaviour
![immagine](https://user-images.githubusercontent.com/4233148/202141976-890b76b4-9a65-4959-8781-326fc9f23287.png)
![immagine](https://user-images.githubusercontent.com/4233148/202142120-9a3e1eac-45cd-4838-a5e2-f3d87551e0bc.png)
so, it's not a derivative constantly flipping between positive and negative values every other steps, as Dymola shows for your MWE and as I expected also OMC to show. This time, OMC is apparentlysimulating "real" behaviour with dozens of steps between the positive and the negative values of the derivative.

I'm still not sure if this is really behaviour that can be explained with the equations (that have lots of suspicious noEvents), or if it is some bug in how discontinuities are handled by the OMC generated code. I'll investigate more and report.

**Comment by casella** [comment]
How do you get the Dymola trajectory? I activated protected variable output but still could not visualize that state

**Comment by casella** [comment]
Anyway, I'm not seeing chattering in Dymola. The number of steps for the whole simulation is significant:
```
   Number of result points                 : 81486
   Number of grid points                   : 81486
   Number of accepted steps                : 81386
   Number of f-evaluations (dynamics)      : 165418
   Number of crossing function evaluations : 86132
   Number of Jacobian-evaluations          : 18959
   Number of model time events             : 400
```
but reasonable, given the kind of transient involved:
![immagine](https://user-images.githubusercontent.com/4233148/202199448-80b69a5b-0484-4b4b-b50e-85724a9b6cac.png)

I analyzed the step-size in Dymola, see plots (abscissa is the step number, ordinate is step length):

![immagine](https://user-images.githubusercontent.com/4233148/202199824-e4cc5db4-a784-40c3-973f-9304a5e198b6.png)
![immagine](https://user-images.githubusercontent.com/4233148/202200091-1780330a-4000-41d9-b656-d96ceab6c5b2.png)
![immagine](https://user-images.githubusercontent.com/4233148/202200284-346521b5-d51e-449b-bcf8-e001eb9c5403.png)
![immagine](https://user-images.githubusercontent.com/4233148/202200388-98131ca5-302f-477a-a382-3e52a7866469.png)
![immagine](https://user-images.githubusercontent.com/4233148/202200480-e6ee0d64-08fc-4293-aa4d-5ab3a965351a.png)

What I see is that there are many events (or noEvent discontinuities?) that bring the stepsize to very small values abruptly, but the it increases back very quickly. In other words, after each discontinuity the derivative is apparently not changing its sign, so the solution can continue forward in a relatively smooth way. There is no endless cycling or chattering, as far as I can see.

The problem is, we have similar behaviour with the Buildings library, so I cannot just dismiss this issue as a modelling problem. For sure this model is numerically tricky, but it shouldn't be chattering.

**Comment by AnHeuermann** [comment]
> How do you get the Dymola trajectory? I activated protected variable output but still could not visualize that state

Yes, it's very annoying. I tried that and the "ignore HideResult" thing in Dymola simulation setup, but it didn't work. So I took the brute force approach and replaced all `HideResult=true` with `HideResult=false` in the BRSL libraries 😆. Good old search and replace.

**Comment by casella** [comment]
Please see #9711, I'd like to have your feedback on that specific issue. I will continue investigating this one. I am more and more convinced that there is something fishy in how we currently handle noEvent stuff.

**Comment by casella** [comment]
OK, so #9711 is now clarified. In order to proceed, getting #9721 done would help a lot. On the other hand, I grep'd all noEvent expressions in the flat model, there are not that many, I'll check all of them ASAP

**Comment by casella** [comment]
Based on my comment in [#9626 ](https://github.com/OpenModelica/OpenModelica/issues/9626#issuecomment-1320669349), I have a possible clue here. 

Consider this simple case of implicit equation: f(x) = 0 with f(x) = c + if x > 0 then 100*x else x. If you draw y = f(x) on a piece of paper, and use the geometrical interpretation of Newton's method, you can easily see that if you use analytic Jacobian (i.e. draw the tangent to the curve), you can solve f(x) = 0 with exactly one iteration, also in the potentially critical case c = 0.

If you instead use a numerical Jacobian, i.e. draw a secant, if the solution is a the discontinuity (hence c = 0) and you take x and x + dx so that they fall on the opposite sides of the discontinuity, the convergence is severely hindered.

Maybe this is what is happening here, this could explain the wildly erratic value of one element of the Jacobian reported [above](https://github.com/OpenModelica/OpenModelica/issues/9131#issuecomment-1299495475). I don't really see dramatic chattering, so I think that all noEvent functions are continuous; however, their derivative aren't, so we get in trouble when computing the solution with numerical Jacobians. In fact, in those cases where we could use --generateAnalyticJacbobian, this problem was solved, a fact that supports this hypothesis.

If this is really the root cause of the problem of this ticket, then the solution would be that every time a condition is some noEvent changes between computing f(x) and computing f(x + dx), then you should take dx with the opposite sign, so as to stay on the same side of the discontinuity.

@AnHeuermann, @kabdelhak what do you think?

**Comment by AnHeuermann** [comment]
I checked most `noEvent` calls (that are no booleans), but could not detect any chattering after time `t=0.5`:

```
LOG_NOEVENT       | info    | At time 0.469507647876833 noEvent condition 3 changed: false
LOG_NOEVENT       | info    | At time 0.500390270738674 noEvent condition 2 changed: true
LOG_NOEVENT       | info    | At time 0.500604401519146 noEvent condition 3 changed: true
LOG_SUCCESS       | info    | The simulation finished successfully.
```

I could have missed some (I did the idea of #9721 partially by hand) and I didn't check noEvent on expressions of type real/integer.

**Comment by AnHeuermann** [comment]
Debugging with @casella and @kabdelhak we found that there is a delay with zero delay time in `Rexroth_Valves.Internal.Templates.Components.ValveDynamics`. Removing `Modelica.Blocks.Nonlinear.FixedDelay fixedDelay` from this example seems to resolve the issue.

So we don't handle zero-delay correctly in this case.

`stopTime=0.6`:
```
timer
    0.233618s          reading init.xml
   0.0166891s [ 39.2%] simulation
   0.0425915s [100.0%] total
   0.0356979s          reading info.xml
   0.0057098s [ 13.4%] pre-initialization
   0.0039696s [  9.3%] initialization
    4.71e-05s [  0.1%] steps
   0.0009559s [  2.2%] solver (excl. callbacks)
   0.0123202s [ 28.9%] creating output-file
   0.0023808s [  5.6%] event-handling
    0.000519s [  1.2%] overhead
events
solver: dassl
  698 steps taken
  966 calls of functionODE
  151 evaluations of jacobian
   34 error test failures
    0 convergence test failures
0.001465s time of jacobian evaluation
The simulation finished successfully.
```

With the delay we have like 20000 steps instead of 698.

**Comment by casella** [comment]
The semantics of `delay(<expr>, delayTime)` when `delayTime = 0` is not very clear in the language specification. I just opened modelica/ModelicaSpecification#3287 on this topic.

While writing that I think I understood what is the root cause of this issue: the block with zero delay is on an important signal path of the model, and I guess that the backend is either not considering that, or considering it in a numerically questionable way when computing the Jacobian of the RHS of the DAEs. This explains why one element of the numerical Jacobian oscillates widely; the position of that element in the Jacobian is also consistent with the location of the zero-delay block in the model.

**Comment by casella** [comment]
#9735 proposes a fix in the NF.

**Comment by casella** [comment]
@niklwors, the test cases you reported above now runs with the following statistics:
```
### STATISTICS ###
timer
    0.251586s          reading init.xml
     300.387s [ 81.6%] simulation
     368.113s [100.0%] total
   0.0390606s          reading info.xml
   0.0184163s [  0.0%] pre-initialization
   0.0041692s [  0.0%] initialization
   0.0001045s [  0.0%] steps
     33.3584s [  9.1%] solver (excl. callbacks)
     1.08479s [  0.3%] creating output-file
     33.2095s [  9.0%] event-handling
   0.0504988s [  0.0%] overhead
events
solver: dassl
16840868 steps taken
52138284 calls of functionODE
25396216 evaluations of jacobian
11799778 error test failures
   25 convergence test failures
216.1s time of jacobian evaluation
```
so we reduced the number of steps by a factor 200 and the number of Jacobian evaluations by a factor 1000. 

Please test as soon as the nightly builds incorporate PR #9740 and reopen if you find the same issue with other models.

**Comment by niklwors** [comment]
Thank you very much @casella @AnHeuermann @kabdelhak this looks very good. I will test it with our test library and let you know.

**Comment by niklwors** [comment]
Thank you very much for your great work and effort. The models now are simulating fast.

**Comment by phannebohm** [comment]
> I guess that the backend is either not considering that, or considering it in a numerically questionable way when computing the Jacobian of the RHS of the DAEs.

I suppose we had the wrong symbolic derivative for the delay operator, see #11975. Maybe that PR has some influence on the BR models.
